### PR TITLE
Algorithm to ensure small healthbar portions are visible

### PR DESCRIPTION
**Pull request by MatApple**
Test Data and Results:
```
{
  className: 'unknown',
  value: 50
},
{
  className: 'healthy',
  value: 47
},
{
  className: 'over-capacity',
  value: 1
},
{
  className: 'unhealthy',
  value: 1
},
{
  className: 'staged',
  value: 1
}
```

```
{
  className: 'unknown',
  value: 99
},
{
  className: 'healthy',
  value: 1
},
{
  className: 'over-capacity',
  value: 0
},
{
  className: 'unhealthy',
  value: 0
},
{
  className: 'staged',
  value: 0
}
```

```
{
  className: 'unknown',
  value: 25
},
{
  className: 'healthy',
  value: 25
},
{
  className: 'over-capacity',
  value: 25
},
{
  className: 'unhealthy',
  value: 20
},
{
  className: 'staged',
  value: 5
}
```

```
{
  className: 'unknown',
  value: 91
},
{
  className: 'healthy',
  value: 6
},
{
  className: 'over-capacity',
  value: 1
},
{
  className: 'unhealthy',
  value: 1
},
{
  className: 'staged',
  value: 1
}
```

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @orlandohohmeier

**Comment by rcorral** [comment]
Awwwsome work!

File: src/js/components/StatusBar.js [hunk 1]
```diff
@@ -11,22 +11,74 @@ class StatusBar extends React.Component {
...
+          // Done with growing this portion. Otherwise it stays on the stack
+          // and will get another % added to it from the largest portion
+          // calculated in the next iteration.
+          indexesLessThanThreshold = indexesLessThanThreshold.slice(1);
+        }
+        // Recurse till we have grown everything above threshold
+        stealPortion();
+      }
+    }
+    stealPortion();
```
Comment (rcorral) [inline src/js/components/StatusBar.js]: What's the purpose of defining this function and calling it right after, wouldn't it just be enough with the if statement?

File: src/js/components/StatusBar.js [hunk 2]
```diff
@@ -11,22 +11,74 @@ class StatusBar extends React.Component {
...
+
+      return relativeSize;
+    });
+
+    function stealPortion() {
+      if (indexesLessThanThreshold.length) {
+        // Try to steal some width from the largest portion
+        // and give it to the first portion in indexesLessThanThreshold
+        maxSize = 0;
+        maxIndex = 0;
```
Comment (rcorral) [inline src/js/components/StatusBar.js]: Can you prefix these with `let` I don't think we need to be using the ones from the closure.

Comment (MatApple) [inline src/js/components/StatusBar.js]: Done!

File: src/js/components/StatusBar.js [hunk 3]
```diff
@@ -11,22 +11,72 @@ class StatusBar extends React.Component {
...
+      let relativeSize = value / max * 100;
+
+      if (relativeSize > 0 && relativeSize < threshold) {
+        indexesLessThanThreshold.push(index);
+      }
+
+      return relativeSize;
+    });
+
+    function stealPortion() {
```
Comment (kennyt) [inline src/js/components/StatusBar.js]: what do you think about having `growRatio` and `threshold` as consts on top of the file, then having `stealPortion` taking in `barSizes` and `indexesLessThanThreshold` as arguments, then we can put this function at the top of the file too

**Comment by kennyt** [comment]
LGTM after my last comment

File: src/js/components/StatusBar.js (hunk 3 above)
Comment (MatApple) [inline src/js/components/StatusBar.js]: Done!

File: src/js/components/StatusBar.js [hunk 4]
```diff
@@ -1,6 +1,43 @@
 import React from 'react';
 import classNames from 'classnames';
 
+// Size of largest portion to give to smallest portion. We have a max of 5
+// status types so we want to keep this growth ratio relatively small.
+const growRatio = 0.07;
+// Grow anything below this threshold by stealing from the biggest portion
+const threshold = 7;
```
Comment (kennyt) [inline src/js/components/StatusBar.js]: #nitnation can we all caps and add `_`s to these?

Comment (MatApple) [inline src/js/components/StatusBar.js]: Done!

**Comment by jfurrow** [comment]
This is sweet!

**Comment by MatApple** [comment]
retest this please

**Pull request merged**